### ISSUE TITLE
Improvements around ReproducibleBatchSampler. 

### ISSUE THREAD

**Issue by amatsukawa** [pytorch/ignite#935]
## 🚀 Feature

IMHO an attempt should be made to not wrap or change the data pipeline objects. Despite the name, this `ReproducibleBatchSampler` seems to be about making datasets *resumable* from the middle of an epoch by skipping the first few examples, rather than being *reproducible* since it doesn't appear to set any seeds on the samplers.

(If this is not the case, I think the point is even stronger. I'd prefer not to have seeds and "reproducibility" introduced into my data pipeline in the background. They are all good things, but not when it happens without my knowledge, or without an option to disable it).

Silently wrapping objects or creating new instances is can introduce unexpected issues, eg:
- point brought up in #812, about side effects.
- `ReproducibleBatchSampler` seems to assume it's wrapping the pytorch `BatchSampler`. For instance, it assumes the `BatchSampler` has a `sampler` instance variable. This this the case for the pytorch class, but not required in general. `BatchSampler` is just a sampler where the `__iter__` returns a batch of indexes.
- `ReproducibleBatchSampler` samples all the indexes first. This isn't necessarily a trivial amount of time or memory to hold l those ints if the dataset is large.

First and foremost, I believe this new behavior should be *very* prominently noted in the documentation, change notes, a warning in the logs, etc. I realized after the fact this is a "note" in the `run` method engine,  but as a user who is upgrading from 0.2.1, I would have had no idea this was happening if it wasn't for my implementation of batch sampler trivially not being compatible. 

Perhaps the behavior should be changed to: if the data loader stack is not holding an instance of `ReproducibleBatchSampler`, then engine simply loads and does nothing with the batches that need to be skipped upon resume. Users who wish to have the `ReproducibleBatchSampler` option can explicitly use this class when constructing `DataLoader`.

Luckily, ignite is a small library and its code is very readable (awesome, very happy user, thanks!). Upon reading the code, I see that I can simply implement my own `ReproducibleBatchSampler` to bypass the three concerns I pointed out above. For the case where batches are simply sampled iid with replacement out of the data, it's sufficient to simply shorten the number of elements of the first call to `__iter__` when the dataset is resumed.

So this is more of a suggestion about default behavior, which I found to be somewhat unexpected.

Thanks for your work on this library!

**Comment by vfdev-5** [comment]
@amatsukawa thanks for the report. Yes, we also agree on that and since v0.4.0 it will be optional. See #895 . Default behaviour will be to run as is without patching the dataloader and without dataflow sync.
Please, take a look at #895 and tell me what do you think about it ?

**Comment by amatsukawa** [comment]
@vfdev-5 thanks for the response. Glad to see this is already being addressed. The solution in #895 seems good to me.

**Comment by amatsukawa** [comment]
I see you've merged in #938, thanks! I was wondering, would you consider doing a 0.3.1 release since it seems this is blocking the workflow of #795? After some exploration, I realized my workflow is blocked also.

It appears this currently does not work: https://pytorch.org/docs/stable/data.html#disable-automatic-batching. When automatic batching is disabled, a sampler that samples a batch of indexes is passed in as the `sampler` argument, not `batch_sampler`. I think solving this is simpler, and simply involves some extra checks to see if `batch_size` is `None` in the loader. But anyhow, it can be side stepped with the new changes to master.

I know I could install from a local clone, but it would be much easier to be able to install from pypi :)

**Comment by vfdev-5** [comment]
> I see you've merged in #938, thanks!

@amatsukawa it is not yet merged to master due to missing updates on examples. It is merged to `new_engine` branch.

Once we merge `new_engine` to `master`, you can install nightly release using pypi as 
```
pip install --pre pytorch-ignite
```

If you are blocked by #795, please use the following temporary solution to keep random states
```python
import torch
import random

def _get_rng_states():
    output = [random.getstate(), torch.get_rng_state()]
    try:
        import numpy as np

        output.append(np.random.get_state())
    except ImportError:
        pass

    return output

def _set_rng_states(rng_states):
    random.setstate(rng_states[0])
    torch.set_rng_state(rng_states[1])
    try:
        import numpy as np

        np.random.set_state(rng_states[2])
    except ImportError:
        pass

def keep_random_state(func: Callable):
    """Helper decorator to keep random state of torch, numpy and random intact
    while executing a function. For more details on usage, please see
    `"Concepts/Dataflow synchronization" <https://pytorch.org/ignite/concepts.html#dataflow-synchronization>`_.
    Args:
        func (callable): function to decorate
    """

    @wraps(func)
    def wrapper(*args, **kwargs):
        rng_states = _get_rng_states()
        func(*args, **kwargs)
        _set_rng_states(rng_states)

    return wrapper
```
Thus, you can decorate evaluation handler with `keep_random_state` and after the handler the training will get back its own random states.


And thanks for mentioning about this feature "automatic batching disable" (forgot about it). Yes, we need to check this too for the deterministic engine...

**Comment by amatsukawa** [comment]
Thanks for the response! I'm actually blocked by not being able to disable automatic batching. Do you have a suggested workaround for that? I'd love to be able to update to take advantage of the new checkpointing features.

Perhaps I could inherit from Engine and just intercept the function here with a different one? https://github.com/pytorch/ignite/blob/012cf38acdf7697b01968b00d488e2cbf8ed1632/ignite/engine/engine.py#L716

**Comment by vfdev-5** [comment]
@amatsukawa we are very sorry for that ! We'll update the library soon with more stable v0.4.0 release.

I think what can be temporary done is to convert torch DataLoader into an iterator and specify `epoch_length` in the run:
```python
trainer.run(map(lambda x: x, data_loader), epoch_length=len(data_loader))
```
Based on #714 (so, probably this work only on nightly release)
